Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,473 @@
+                significand.lo += UInt64(digit)
+            }
+        }
+
+        if !hi_digits.isEmpty {
+            significand.hi = UInt64(hi_digits[0])
+            for digit in hi_digits[1...] {
+                significand.hi *= 10
+                significand.hi += UInt64(digit)
+            }
+        }
+
+        var product = UInt128.multiply(significand.hi, by: 100_000_000_000_000_000)
+        product.lo += significand.lo
+
+        if product.lo < significand.lo {
+            product.hi += 1
+        }
+
+        let biased_exponent = exponent + BSONDecimal128.EXPONENT_BIAS
+
+        self.value = UInt128()
+
+        // Encode combination, exponent, and significand.
+        if (product.hi >> 49) & 1 == 1 {
+            // Encode '11' into bits 1 to 3
+            self.value.hi |= (0b11 << 61)
+            self.value.hi |= UInt64(biased_exponent & 0x3FFF) << 47
+            self.value.hi |= product.hi & 0x7FFF_FFFF_FFFF
+        } else {
+            self.value.hi |= UInt64(biased_exponent & 0x3FFF) << 49
+            self.value.hi |= product.hi & 0x1_FFFF_FFFF_FFFF
+        }
+
+        self.value.lo = product.lo
+
+        if sign < 0 {
+            self.value.hi |= 0x8000_0000_0000_0000
+        }
+    }
+
+    private static func convertToDigitsArray(_ decimalString: String) throws -> [UInt8] {
+        var leadingZero = true
+        var digits: [UInt8] = []
+
+        let asciiZero = Character("0").asciiValue ?? 48
+        let asciiNine = Character("9").asciiValue ?? 57
+        let asciiPoint = Character(".").asciiValue ?? 46
+
+        let digitsFromRepr = [UInt8](decimalString.utf8)
+        if digitsFromRepr.count > 1 {
+            for digit in digitsFromRepr {
+                if digit == asciiPoint {
+                    continue
+                }
+                if (digit < asciiZero) && (digit > asciiNine) {
+                    throw BSONError.InvalidArgumentError(message: "Syntax Error: \(digit) is not a digit 0-9")
+                }
+                if digit == asciiZero && leadingZero {
+                    continue
+                }
+                if digit != asciiZero && leadingZero {
+                    // seen a non zero digit
+                    leadingZero = false
+                }
+                digits.append(digit - asciiZero)
+            }
+        } else {
+            digits.append(digitsFromRepr[0] - asciiZero)
+        }
+        return digits
+    }
+
+    private func toString() -> String {
+        // swiftlint:disable:previous cyclomatic_complexity
+        /**
+         * BSON_DECIMAL128_STRING:
+         *
+         * The length of a decimal128 string.
+         *
+         * 1  for the sign
+         * 35 for digits and radix
+         * 2  for exponent indicator and sign
+         * 4  for exponent digits
+         * BSON_DECIMAL128_STRING 42
+         */
+        var exponent: Int
+        var sig_prefix: Int
+
+        let COMBINATION_INFINITY = 30
```

Comment:
some explanation of these constants would be nice, also specifying them as `static let`